src/python/pants/backend/shell/target_types.py

```diff
@@ -30,14 +30,19 @@
 from pants.util.enums import match
 
 
-class ShellSources(Sources):
+class ShellSourceField(Sources):
     # Normally, we would add `expected_file_extensions = ('.sh',)`, but Bash scripts don't need a
     # file extension, so we don't use this.
     uses_source_roots = False
+    expected_num_files = 1
+
+
+class ShellGeneratingSources(Sources):
+    uses_source_roots = False
```

Comment:
NB that the Target API needs to view these sources as different than the file-based targets for all our rules to do the right thing. This split is what unlocks all the PR's benefits.

I anticipate that this will soon subclass something like `GeneratingSources`, rather than `Sources`. See https://github.com/pantsbuild/pants/issues/12953 for that idea.

Reply:
It seems a bit premature to do this until we know what https://docs.google.com/document/d/17XvhHlx8jBn4K9qFRZPe7qaKJQjTfiKMKYRubnzWxjA/edit will look like.

In particular: an unnamed target never actually needs to exist as a Target object if there is a shim for it during BUILD file parsing. And for a named Target generator, after BUILD file parsing, it shouldn't need a Sources field because it will be sources-less, with dependencies on its generated Targets instead.

Reply:
To be clear, the idea of GeneratingSources is that it's not normal Sources. It's not representing ownership of the files, only a relationship that those files trigger generation. 

Either way, this PR improves the modeling by having the generator sources be a different subclass than the atom sources.